circus-train-gcp/src/main/java/com/hotels/bdp/circustrain/gcp/GCPCredentialCopier.java

```diff
@@ -15,77 +15,107 @@
+    if (security == null || isBlank(security.getCredentialProvider())) {
+      throw new IllegalArgumentException("gcp-security credential-provider must be set");
+    }
+    this.security = security;
+    credentialsFileRelativePath = getCredentialProviderRelativePath();
+    hdfsGsCredentialDirectory = getHdfsGsCredentialDirectory();
+    hdfsGsCredentialAbsolutePath = getHdfsGsCredentialAbsolutePath();
+    LOG.debug("Credential Provider URI = {}", credentialsFileRelativePath);
+    LOG.debug("Temporary HDFS Google Cloud credential location set to {}", hdfsGsCredentialDirectory);
+    LOG.debug("HDFS Google Cloud credential path will be {}", hdfsGsCredentialAbsolutePath);
+  }
+
+  private String getHdfsGsCredentialDirectory() {
+    if (isBlank(security.getDistributedFileSystemWorkingDirectory())) {
+      return DEFAULT_HDFS_PREFIX + RANDOM_STRING;
+    } else {
+      return security.getDistributedFileSystemWorkingDirectory();
     }
-    this.credentialProvider = credentialProvider;
+  }
+
+  private String getCredentialProviderRelativePath() {
+    // The DistributedCache of Hadoop can only link files from their relative path to the working directory
+    java.nio.file.Path currentDirectory = Paths.get(WORKING_DIRECTORY);
+    java.nio.file.Path absolutePathToCredentialsFile = Paths.get(security.getCredentialProvider());
+    java.nio.file.Path pathRelative = currentDirectory.relativize(absolutePathToCredentialsFile);
+    return pathRelative.toString();
+  }
+
+  private String getHdfsGsCredentialAbsolutePath() {
+    return hdfsGsCredentialDirectory + CACHED_CREDENTIAL_NAME;
   }
 
   void copyCredentials() {
     try {
-      copyCredentialIntoWorkingDirectory();
       copyCredentialIntoHdfs();
-      copyCredentialIntoDistributedCache();
+      linkRelativePathInDistributedCache();
     } catch (IOException | URISyntaxException e) {
       throw new CircusTrainException(e);
     }
   }
 
-  private void copyCredentialIntoWorkingDirectory() throws IOException {
-    File source = new File(credentialProvider);
-    File destination = new File(WORKING_DIRECTORY + CACHED_CREDENTIAL_NAME);
-    destination.deleteOnExit();
-    LOG.debug("Copying credential into working directory {}", destination);
-    FileUtils.copyFile(source, destination);
-  }
-
   private void copyCredentialIntoHdfs() throws IOException {
-    Path source = new Path(credentialProvider);
-    Path destination = new Path(HDFS_GS_CREDENTIAL_ABSOLUTE_PATH);
-    Path destinationFolder = new Path(HDFS_GS_CREDENTIAL_DIRECTORY);
+    /*
+     * The Google credentials file must be present in HDFS so that the DistCP map reduce job can access it upon
+     * replication.
+     */
+    Path source = new Path(credentialsFileRelativePath);
+    Path destination = new Path(hdfsGsCredentialAbsolutePath);
+    Path destinationFolder = new Path(hdfsGsCredentialDirectory);
     fs.deleteOnExit(destinationFolder);
```

Comment:
This is not tested, assert it in the test I suggested.

The destinationFolder looks wrong, it could be the working directory that a user configures `security.getDistributedFileSystemWorkingDirectory()`, which you will delete. What if this working directory is used by other jobs you'll be deleting their data. 
Only delete folders/files you create.